File: src/anylayout/extras/SizeCalculatorUtility.java
package anylayout.extras;

import anylayout.SizeCalculator;
import fj.F;

import java.awt.Component;

public final class SizeCalculatorUtility
{
	public static SizeCalculator absoluteSize(final int width,final int height)
	{
		return new SizeCalculator()
		{
			public int getWidth()
			{
				return width;
			}

			public int getHeight()
			{
				return height;
			}
		};
	}

	public static F<Component,Integer> getPreferredWidth()
	{
		return new F<Component,Integer>()
		{
			public Integer f(final Component component)
                        {
				return component.getPreferredSize().width;
		        }
		};
	}

	public static F<Component,Integer> getPreferredHeight()
	{
		return new F<Component,Integer>()
		{
			public Integer f(final Component component)
			{
				return component.getPreferredSize().height;
			}
		};
	}

	public static SizeCalculator constrain(final SizeCalculator sizeCalculator,final int maxWidth,final int maxHeight)
        {
		return new SizeCalculator()
		{
			public int getHeight()
                        {
				return Math.min(maxHeight,sizeCalculator.getHeight());
                        }

			public int getWidth()
                        {
				return Math.min(maxWidth,sizeCalculator.getWidth());
                        }			
		};
        }
}

File: src/anylayout/extras/SizeCalculatorUtility.scala
package anylayout.extras

import anylayout.SizeCalculator
import java.awt.Component

object SizeCalculatorUtility {
 case class absoluteSize(override val getWidth: Int, override val getHeight: Int) extends SizeCalculator
 val getPreferredWidth: Component => Int = _.getPreferredSize.width
 val getPreferredHeight: Component => Int = _.getPreferredSize.height
 def constrain(sizeCalculator: SizeCalculator, maxWidth: Int, maxHeight: Int) = new SizeCalculator {
  def getHeight = Math.min(maxHeight, sizeCalculator.getHeight)
  def getWidth = Math.min(maxWidth, sizeCalculator.getWidth)
 }
}
